kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/PartitionLog.java

```diff
@@ -0,0 +1,332 @@
+import io.streamnative.pulsar.handlers.kop.format.EntryFormatter;
+import io.streamnative.pulsar.handlers.kop.format.KafkaMixedEntryFormatter;
+import io.streamnative.pulsar.handlers.kop.utils.MessageMetadataUtils;
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.bookkeeper.mledger.ManagedLedger;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.errors.CorruptRecordException;
+import org.apache.kafka.common.errors.RecordTooLargeException;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.record.InvalidRecordException;
+import org.apache.kafka.common.record.MemoryRecords;
+import org.apache.kafka.common.record.MutableRecordBatch;
+import org.apache.kafka.common.record.RecordBatch;
+import org.apache.kafka.common.utils.Time;
+import org.apache.pulsar.broker.service.persistent.PersistentTopic;
+import org.apache.pulsar.common.naming.TopicName;
+
+/**
+ * An append-only log for storing messages. Mapping to Kafka Log.scala.
+ */
+@Slf4j
+@AllArgsConstructor
+public class PartitionLog {
+    private KafkaServiceConfiguration kafkaConfig;
+    private Time time;
+    private TopicPartition topicPartition;
+    private String namespacePrefix;
+    private String fullPartitionName;
+    private EntryFormatter entryFormatter;
+    private Optional<TransactionCoordinator> transactionCoordinator;
+
+
+    // A lock that guards all modifications to the log
+    private final Object lock = new Object();
+
+    @Data
+    @AllArgsConstructor
+    public static final class LogAppendInfo {
+        private Optional<Long> firstOffset;
+        private Long lastOffset;
+        private Integer shallowCount;
+        private Boolean offsetsMonotonic;
+        private Long lastOffsetOfFirstBatch;
+        private Integer validBytes;
+
+        public Long numMessages() {
+            return firstOffset.map(firstOffsetVal -> {
+                if (firstOffsetVal >= 0 && lastOffset >= 0) {
+                    return lastOffset - firstOffsetVal + 1;
+                }
+                return 0L;
+            }).orElse(0L);
+        }
+    }
+
+    public CompletableFuture<Long> appendRecords(final MemoryRecords records,
+                                                 final short version,
+                                                 final AppendRecordsContext appendRecordsContext) {
+        return append(records, version, false, appendRecordsContext);
+    }
+
+    /**
+     * Append this message to pulsar.
+     *
+     * This method will generally be responsible for assigning offsets to the messages,
+     * however if the assignOffsets=false flag is passed we will only check that the existing offsets are valid.
```

Comment:
It looks like these two lines were copied from Kafka? Here it's a little different in KoP. We don't have `assignOffsets=false` flag. Instead, we only allows `MixedKafkaEntryFormatter` to assign offsets and the original `KafkaV1EntryFormatter` (it actually also supports V0) doesn't need to assign offsets.

Reply:
Addressed.